### DISCUSSION TITLE
Improving incidence2 objects

### DISCUSSION THREAD

**Discussion by TimTaylor** [reconhub/incidence2#63] [Ideas]
There are a few areas where I think we can improve `incidence2`:
* Move all interval implementation details in to grates (currently we have a hybrid approach).
* Remove the cumulative attribute. It adds complexity with little gain and we probably want to discourage it's use anyway.
* Formalise how we work with additional variables that are not grouping variables (e.g. you may want a day of the week variable).
* Improve plotting implementation details.
* Formalise the invariant checks we do (currently within dplyr.R)

Ignoring tibble subclass for time being roughly speaking the structure of an incidence2 object could be:
```
structure(x, date_index , groups, counts, other_variables)
``` 
and the invariants would be:
* No duplicate rows
* Exactly one date_index
* 0 or more groups
* 1 or more counts
* 0 or more other variables

**Comment by TimTaylor** [comment]
@thibautjombart for awareness

**Reply by thibautjombart** [reply]
sounds great!

**Comment by TimTaylor** [comment]
invariants could be checked by a `validate_incidence()` function.  You could then just build methods for data.frames and the incidence2 methods will be a safely wrapper lapply over the top of this.

**Reply by seabbs** [reply]
Really like this idea!

**Comment by seabbs** [comment]
Big fan of all of these ideas.

**Comment by TimTaylor** [comment]
Cool. 

> * Remove the cumulative attribute. It adds complexity with little gain and we probably want to discourage it's use anyway.

This can be done easily.

> * Move all interval implementation details in to grates (currently we have a hybrid approach).

This is less trivial and annoyingly the other improvements will depend on this - so this will be the priority when I'm next on this.

>  * Formalise how we work with additional variables that are not grouping variables (e.g. you may want a day of the week variable).

Once the additional grates work is done and we've formalised the invariants check then this should follow easily

> * Improve plotting implementation details.

This code is currently a bit fiddly so will try and improve once the things above are dealt with.  Relies a lot on getting the scales right in grates.

**Comment by TimTaylor** [comment]
A quick update.  I think this is now resonably well implemented. https://github.com/reconverse/incidence2/commit/368b42f047afeaef19abd324670defe418efee8b was the final bit to fall in to place.

There is now a minimal constructor (`new_incidence()`) and validator `validate_incidence()` which can be easily used to create an incidence object from a data.frame and check that the invariants hold.  I've also added a new function `as_incidence()` which allows users to specify the function they want to apply to grates rather than forcing them to use the grates implementation.  This will be really useful if the `clock` package implements decent scales for ggplot. The `incidence()` function remains the "user-friendly" function that makes use of `grates` under the hood. I've managed to do this without breaking changes which is cool.

Will leave this to sit in the master branch for a good few weeks before pushing to CRAN but am hoping the package is now close to entering "maintenance" mode ;)

**Reply by seabbs** [reply]
This is all looking really nice Tim!
